pkg/istiodcert/worker.go

```diff
@@ -0,0 +1,277 @@
+func (dicp *DynamicIstiodCertProvisioner) AddControllersToManager(mgr manager.Manager) error {
+	b := ctrl.NewControllerManagedBy(mgr)
+
+	b.For(
+		new(cmapi.Certificate), builder.WithPredicates(predicate.NewPredicateFuncs(func(obj client.Object) bool {
+			// Only process one specific cert which was requested
+			return obj.GetName() == dicp.opts.CertificateName && obj.GetNamespace() == dicp.opts.CertificateNamespace
+		})))
+
+	// when the issuer changes, trigger a re-reconciliation
+	b.WatchesRawSource(source.Channel(dicp.reconcileChan, handler.EnqueueRequestsFromMapFunc(
+		func(context.Context, client.Object) []reconcile.Request {
+			return []reconcile.Request{ctrl.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      dicp.opts.CertificateName,
+					Namespace: dicp.opts.CertificateNamespace,
+				},
+			}}
+		})))
+
+	err := b.Complete(dicp)
+	if err != nil {
+		return err
+	}
+
+	err = mgr.Add(dicp)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (dicp *DynamicIstiodCertProvisioner) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	dicp.issuerRefMutex.Lock()
+	defer dicp.issuerRefMutex.Unlock()
+
+	if dicp.issuerRef == nil {
+		dicp.log.Info("exiting reconcile of dynamic istiod early; no issuerRef is set")
+		return ctrl.Result{}, nil
+	}
+
+	dicp.log.Info("reconciling dynamic istiod cert", "issuer-name", dicp.issuerRef.Name, "issuer-kind", dicp.issuerRef.Kind, "issuer-group", dicp.issuerRef.Group)
+
+	spiffeID := fmt.Sprintf("spiffe://%s/ns/%s/sa/istiod-service-account", dicp.trustDomain, req.Namespace)
+
+	commonName, dnsNames := makeDNSNamesFromRevisions(req.Namespace, dicp.opts.IstioRevisions)
+
+	if len(dicp.opts.AdditionalDNSNames) > 0 {
+		dnsNames = append(dnsNames, dicp.opts.AdditionalDNSNames...)
+	}
+
+	desiredSpec := cmapi.CertificateSpec{
+		CommonName:  commonName,
+		DNSNames:    dnsNames,
+		URIs:        []string{spiffeID},
+		SecretName:  "istiod-tls",
+		Duration:    &metav1.Duration{Duration: dicp.opts.Duration},
+		RenewBefore: &metav1.Duration{Duration: dicp.opts.RenewBefore},
+		PrivateKey: &cmapi.CertificatePrivateKey{
+			RotationPolicy: cmapi.RotationPolicyAlways,
+			Algorithm:      dicp.opts.CMKeyAlgorithm,
+			Size:           dicp.opts.KeySize,
+		},
+		RevisionHistoryLimit: ptr.To(int32(1)),
+		IssuerRef:            *dicp.issuerRef,
+	}
+
+	cert, err := dicp.certManagerClient.Get(ctx, req.Name, metav1.GetOptions{})
+
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, fmt.Errorf("failed to fetch cert: %s", err)
+		}
+
+		cert := cmapi.Certificate{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      req.Name,
+				Namespace: req.Namespace,
+			},
+			Spec: desiredSpec,
+		}
+
+		_, err = dicp.certManagerClient.Create(ctx, &cert, metav1.CreateOptions{})
+		return ctrl.Result{}, err
+	}
+
+	cert.Spec = desiredSpec
+
+	_, err = dicp.certManagerClient.Update(ctx, cert, metav1.UpdateOptions{})
```

Comment:
There doesn't appear to be an endless loop from testing this; by running the commands in the description of this PR I see the below logs when I update the configmap to trigger a reconciliation.

There aren't any further reconciles after that.

I'm open to changing this and adding a `needsUpdate` var or something but there weren't any issues I saw from unconditionally updating here, presumably because controller-runtime is doing something smart. What do you think?

```json
{
  "ts": 1722337665338.1196,
  "logger": "cert-manager.runtime-config-watcher",
  "caller": "certmanager/certmanager.go:343",
  "msg": "Changed active issuerRef in response to runtime configuration ConfigMap",
  "config-map-name": "runtime-config-map",
  "config-map-namespace": "cert-manager",
  "v": 0,
  "issuer-name": "runtimeconfig-ca-issuer",
  "issuer-kind": "Issuer",
  "issuer-group": "cert-manager.io"
}
{
  "ts": 1722337665338.2039,
  "logger": "istiod-dynamic",
  "caller": "istiodcert/worker.go:128",
  "msg": "triggering reconciliation of istiod cert after issuer change",
  "v": 0,
  "cert_name": "istiod-dynamic",
  "cert_namespace": "istio-system"
}
{
  "ts": 1722337665338.4597,
  "logger": "manager",
  "caller": "controller/controller.go:310",
  "msg": "Reconciling",
  "controller": "certificate",
  "controllerGroup": "cert-manager.io",
  "controllerKind": "Certificate",
  "Certificate": {
    "name": "istiod-dynamic",
    "namespace": "istio-system"
  },
  "namespace": "istio-system",
  "name": "istiod-dynamic",
  "reconcileID": "abdc97c7-99fa-4737-9c35-2e94e4066d3a",
  "v": 5
}
{
  "ts": 1722337665338.4792,
  "logger": "istiod-dynamic",
  "caller": "istiodcert/worker.go:185",
  "msg": "reconciling dynamic istiod cert",
  "v": 0,
  "issuer-name": "runtimeconfig-ca-issuer",
  "issuer-kind": "Issuer",
  "issuer-group": "cert-manager.io"
}
{
  "ts": 1722337665346.3403,
  "logger": "manager",
  "caller": "controller/controller.go:339",
  "msg": "Reconcile successful",
  "controller": "certificate",
  "controllerGroup": "cert-manager.io",
  "controllerKind": "Certificate",
  "Certificate": {
    "name": "istiod-dynamic",
    "namespace": "istio-system"
  },
  "namespace": "istio-system",
  "name": "istiod-dynamic",
  "reconcileID": "abdc97c7-99fa-4737-9c35-2e94e4066d3a",
  "v": 5
}
{
  "ts": 1722337665346.528,
  "logger": "manager",
  "caller": "controller/controller.go:310",
  "msg": "Reconciling",
  "controller": "certificate",
  "controllerGroup": "cert-manager.io",
  "controllerKind": "Certificate",
  "Certificate": {
    "name": "istiod-dynamic",
    "namespace": "istio-system"
  },
  "namespace": "istio-system",
  "name": "istiod-dynamic",
  "reconcileID": "6930cebc-1625-4f8b-a9c0-5f62fddd30a7",
  "v": 5
}
{
  "ts": 1722337665346.566,
  "logger": "istiod-dynamic",
  "caller": "istiodcert/worker.go:185",
  "msg": "reconciling dynamic istiod cert",
  "v": 0,
  "issuer-name": "runtimeconfig-ca-issuer",
  "issuer-kind": "Issuer",
  "issuer-group": "cert-manager.io"
}
{
  "ts": 1722337665350.698,
  "logger": "manager",
  "caller": "controller/controller.go:324",
  "msg": "Reconciler error",
  "controller": "certificate",
  "controllerGroup": "cert-manager.io",
  "controllerKind": "Certificate",
  "Certificate": {
    "name": "istiod-dynamic",
    "namespace": "istio-system"
  },
  "namespace": "istio-system",
  "name": "istiod-dynamic",
  "reconcileID": "6930cebc-1625-4f8b-a9c0-5f62fddd30a7",
  "err": "Operation cannot be fulfilled on certificates.cert-manager.io \"istiod-dynamic\": the object has been modified; please apply your changes to the latest version and try again"
}
{
  "ts": 1722337665350.729,
  "logger": "manager",
  "caller": "controller/controller.go:310",
  "msg": "Reconciling",
  "controller": "certificate",
  "controllerGroup": "cert-manager.io",
  "controllerKind": "Certificate",
  "Certificate": {
    "name": "istiod-dynamic",
    "namespace": "istio-system"
  },
  "namespace": "istio-system",
  "name": "istiod-dynamic",
  "reconcileID": "0c0fb9f1-b5dd-408f-8e2b-43cbbd3130a8",
  "v": 5
}
{
  "ts": 1722337665350.734,
  "logger": "istiod-dynamic",
  "caller": "istiodcert/worker.go:185",
  "msg": "reconciling dynamic istiod cert",
  "v": 0,
  "issuer-name": "runtimeconfig-ca-issuer",
  "issuer-kind": "Issuer",
  "issuer-group": "cert-manager.io"
}
{
  "ts": 1722337665355.499,
  "logger": "manager",
  "caller": "controller/controller.go:324",
  "msg": "Reconciler error",
  "controller": "certificate",
  "controllerGroup": "cert-manager.io",
  "controllerKind": "Certificate",
  "Certificate": {
    "name": "istiod-dynamic",
    "namespace": "istio-system"
  },
  "namespace": "istio-system",
  "name": "istiod-dynamic",
  "reconcileID": "0c0fb9f1-b5dd-408f-8e2b-43cbbd3130a8",
  "err": "Operation cannot be fulfilled on certificates.cert-manager.io \"istiod-dynamic\": the object has been modified; please apply your changes to the latest version and try again"
}
{
  "ts": 1722337665355.532,
  "logger": "manager",
  "caller": "controller/controller.go:310",
  "msg": "Reconciling",
  "controller": "certificate",
  "controllerGroup": "cert-manager.io",
  "controllerKind": "Certificate",
  "Certificate": {
    "name": "istiod-dynamic",
    "namespace": "istio-system"
  },
  "namespace": "istio-system",
  "name": "istiod-dynamic",
  "reconcileID": "7f6aa819-e86c-426e-9a43-cca148c75d08",
  "v": 5
}
{
  "ts": 1722337665355.541,
  "logger": "istiod-dynamic",
  "caller": "istiodcert/worker.go:185",
  "msg": "reconciling dynamic istiod cert",
  "v": 0,
  "issuer-name": "runtimeconfig-ca-issuer",
  "issuer-kind": "Issuer",
  "issuer-group": "cert-manager.io"
}
{
  "ts": 1722337665359.5068,
  "logger": "manager",
  "caller": "controller/controller.go:339",
  "msg": "Reconcile successful",
  "controller": "certificate",
  "controllerGroup": "cert-manager.io",
  "controllerKind": "Certificate",
  "Certificate": {
    "name": "istiod-dynamic",
    "namespace": "istio-system"
  },
  "namespace": "istio-system",
  "name": "istiod-dynamic",
  "reconcileID": "7f6aa819-e86c-426e-9a43-cca148c75d08",
  "v": 5
}
{
  "ts": 1722337665359.5374,
  "logger": "manager",
  "caller": "controller/controller.go:310",
  "msg": "Reconciling",
  "controller": "certificate",
  "controllerGroup": "cert-manager.io",
  "controllerKind": "Certificate",
  "Certificate": {
    "name": "istiod-dynamic",
    "namespace": "istio-system"
  },
  "namespace": "istio-system",
  "name": "istiod-dynamic",
  "reconcileID": "59ee22df-1480-4d77-ab70-0dd1b236e655",
  "v": 5
}
{
  "ts": 1722337665359.542,
  "logger": "istiod-dynamic",
  "caller": "istiodcert/worker.go:185",
  "msg": "reconciling dynamic istiod cert",
  "v": 0,
  "issuer-name": "runtimeconfig-ca-issuer",
  "issuer-kind": "Issuer",
  "issuer-group": "cert-manager.io"
}
{
  "ts": 1722337665363.4814,
  "logger": "manager",
  "caller": "controller/controller.go:339",
  "msg": "Reconcile successful",
  "controller": "certificate",
  "controllerGroup": "cert-manager.io",
  "controllerKind": "Certificate",
  "Certificate": {
    "name": "istiod-dynamic",
    "namespace": "istio-system"
  },
  "namespace": "istio-system",
  "name": "istiod-dynamic",
  "reconcileID": "59ee22df-1480-4d77-ab70-0dd1b236e655",
  "v": 5
}
{
  "ts": 1722337665428.861,
  "logger": "manager",
  "caller": "controller/controller.go:310",
  "msg": "Reconciling",
  "controller": "certificate",
  "controllerGroup": "cert-manager.io",
  "controllerKind": "Certificate",
  "Certificate": {
    "name": "istiod-dynamic",
    "namespace": "istio-system"
  },
  "namespace": "istio-system",
  "name": "istiod-dynamic",
  "reconcileID": "3d8d7d75-bc09-49c2-a860-7e1a4cfc8e4e",
  "v": 5
}
{
  "ts": 1722337665428.881,
  "logger": "istiod-dynamic",
  "caller": "istiodcert/worker.go:185",
  "msg": "reconciling dynamic istiod cert",
  "v": 0,
  "issuer-name": "runtimeconfig-ca-issuer",
  "issuer-kind": "Issuer",
  "issuer-group": "cert-manager.io"
}
{
  "ts": 1722337665538.9067,
  "caller": "rest/request.go:629",
  "msg": "Waited for 109.967334ms due to client-side throttling, not priority and fairness, request: GET:https://10.0.0.1:443/apis/cert-manager.io/v1/namespaces/istio-system/certificates/istiod-dynamic",
  "v": 3
}
{
  "ts": 1722337665739.4514,
  "caller": "rest/request.go:629",
  "msg": "Waited for 199.192583ms due to client-side throttling, not priority and fairness, request: PUT:https://10.0.0.1:443/apis/cert-manager.io/v1/namespaces/istio-system/certificates/istiod-dynamic",
  "v": 3
}
{
  "ts": 1722337665749.1682,
  "logger": "manager",
  "caller": "controller/controller.go:339",
  "msg": "Reconcile successful",
  "controller": "certificate",
  "controllerGroup": "cert-manager.io",
  "controllerKind": "Certificate",
  "Certificate": {
    "name": "istiod-dynamic",
    "namespace": "istio-system"
  },
  "namespace": "istio-system",
  "name": "istiod-dynamic",
  "reconcileID": "3d8d7d75-bc09-49c2-a860-7e1a4cfc8e4e",
  "v": 5
}
{
  "ts": 1722337665749.242,
  "logger": "manager",
  "caller": "controller/controller.go:310",
  "msg": "Reconciling",
  "controller": "certificate",
  "controllerGroup": "cert-manager.io",
  "controllerKind": "Certificate",
  "Certificate": {
    "name": "istiod-dynamic",
    "namespace": "istio-system"
  },
  "namespace": "istio-system",
  "name": "istiod-dynamic",
  "reconcileID": "82531b83-df27-482d-aaa3-001249f64752",
  "v": 5
}
{
  "ts": 1722337665749.2615,
  "logger": "istiod-dynamic",
  "caller": "istiodcert/worker.go:185",
  "msg": "reconciling dynamic istiod cert",
  "v": 0,
  "issuer-name": "runtimeconfig-ca-issuer",
  "issuer-kind": "Issuer",
  "issuer-group": "cert-manager.io"
}
{
  "ts": 1722337665939.9775,
  "caller": "rest/request.go:629",
  "msg": "Waited for 190.461ms due to client-side throttling, not priority and fairness, request: GET:https://10.0.0.1:443/apis/cert-manager.io/v1/namespaces/istio-system/certificates/istiod-dynamic",
  "v": 3
}
{
  "ts": 1722337666139.1604,
  "caller": "rest/request.go:629",
  "msg": "Waited for 193.303292ms due to client-side throttling, not priority and fairness, request: PUT:https://10.0.0.1:443/apis/cert-manager.io/v1/namespaces/istio-system/certificates/istiod-dynamic",
  "v": 3
}
{
  "ts": 1722337666149.929,
  "logger": "manager",
  "caller": "controller/controller.go:339",
  "msg": "Reconcile successful",
  "controller": "certificate",
  "controllerGroup": "cert-manager.io",
  "controllerKind": "Certificate",
  "Certificate": {
    "name": "istiod-dynamic",
    "namespace": "istio-system"
  },
  "namespace": "istio-system",
  "name": "istiod-dynamic",
  "reconcileID": "82531b83-df27-482d-aaa3-001249f64752",
  "v": 5
}
```